setup.py

```diff
@@ -1,95 +1,108 @@
-    extras_require = {"cpu": ["tensorflow==" + tf_version], "gpu": ["tensorflow-gpu==" + tf_version]}
-else:
-    extras_require = {"cpu": ["tensorflow-cpu==" + tf_version], "gpu": ["tensorflow==" + tf_version]}
-tf_spec = importlib.util.find_spec("tensorflow")
-if tf_spec:
-    tf_install_dir = tf_spec.submodule_search_locations[0]
+if tf_version in SpecifierSet("<1.15") or tf_version in SpecifierSet(">=2.0,<2.1"):
+    extras_require = {
+        "cpu": [f"tensorflow=={tf_version}"],
+        "gpu": [f"tensorflow-gpu=={tf_version}"],
+    }
 else:
-    site_packages_path = path.join(path.dirname(path.__file__), 'site-packages')
-    tf_spec = importlib.machinery.FileFinder(site_packages_path).find_spec("tensorflow")
-    if tf_spec:
-        tf_install_dir = tf_spec.submodule_search_locations[0]
-    else:
-        setup_requires.append("tensorflow==" + tf_version)
-        tf_install_dir = path.join(path.dirname(path.abspath(__file__)), '.egg',
-                                   pkg_resources.Distribution(project_name="tensorflow", version=tf_version,
-                                                              platform=get_platform()).egg_name(),
-                                   'tensorflow')
+    extras_require = {
+        "cpu": [f"tensorflow-cpu=={tf_version}"],
+        "gpu": [f"tensorflow=={tf_version}"],
+    }
 
-# add cmake as a build requirement if cmake>3.7 is not installed
+# get tensorflow spec
+tf_spec = find_spec("tensorflow")
+if not tf_spec:
+    # purelib gets site-packages path
+    site_packages = get_path("purelib")
+    if site_packages:
+        tf_spec = FileFinder(site_packages).find_spec("tensorflow")
+
+# get install dir from spec
 try:
-    if LegacyVersion(get_cmake_version()) < LegacyVersion("3.7"):
-        setup_requires.append('cmake')
-except SKBuildError:
-    setup_requires.append('cmake')
+    tf_install_dir = tf_spec.submodule_search_locations[0]  # type: ignore
+    # AttributeError if ft_spec is None
+    # TypeError if submodule_search_locations are None
+    # IndexError if submodule_search_locations is an empty list
+except (AttributeError, TypeError, IndexError):
+    setup_requires.append(f"tensorflow=={tf_version}")
+    dist = Distribution(
+        project_name="tensorflow", version=tf_version, platform=get_platform()
+    ).egg_name()
+    tf_install_dir = Path(__file__).parent.resolve().joinpath(".egg", dist, "tensorflow").resolve()
 
+# add cmake as a build requirement if cmake>3.7 is not installed
 try:
-    makedirs('deepmd')
-except OSError:
-    pass
+    cmake_version = get_cmake_version()
+except SKBuildError:
+    setup_requires.append("cmake")
+else:
+    if cmake_version in SpecifierSet("<3.7"):
+        setup_requires.append("cmake")
 
+Path("deepmd").mkdir(exist_ok=True)
 
 setup(
     name="deepmd-kit",
-    setup_requires=setup_requires,
-    use_scm_version={'write_to': 'deepmd/_version.py'},
+    setup_requires=INSTALL_REQUIRES,
```

Comment:
maybe broken here...
```suggestion
    setup_requires=setup_requires,
```